libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL-fp.cpp

```diff
@@ -0,0 +1,36 @@
+
+#include <WiFiClientSecureBearSSL.h>
+
+namespace BearSSL {
+
+static uint8_t htoi (unsigned char c)
+{
+    return c >= '0' && c <= '9'? c - '0'
+         : c >= 'A' && c <= 'F'? c - 'A' + 10
+         : c >= 'a' && c <= 'f'? c - 'a' + 10
+         : 255;
+}
+
+void WiFiClientSecure::setFingerprint(const String& fingerprint) {
+
+    uint8_t fp [20];
+    uint8_t c, d;
+    int idx = 0;
+    const char* _fingerprint = fingerprint.c_str();
+
+    while (idx < 20 && (c = *_fingerprint++) && (d = *_fingerprint++)) {
+        c = htoi(c);
+        d = htoi(d);
+        if (c > 15 || d > 15) {
```

Comment:
Suggest the following to be more pedantic about verification (forgive no indenting, doing this in the review window in proportional font):
````
while (idx < 20 && (c = pgm_read_byte(_fingerprint++)) && (d = pgm_read_byte(_fingerprint++))) { 
  c = htoi(c);
  d = htoi(d);
  // Skip 0 or more spaces for all but last entry (could drop while and only check for 1 ' ' as well)
  while ((idx < 19) && pgm_read_byte(_fingerprint) && (pgm_read_byte(_fingerprint)==' ')) { _fingerprint++; }
  if ((c>15) || (d>15)) { return false; }
  fp[idx++] = (c<<4)|d;
}
if ((idx != 20) || pgm_read_byte(_fingerprint)) { return false; } // Some garbage left over at end of line
return setFingerprint(fp);
````

Reply:
With this code fingerprint *must* be in PROGMEM right ?

Reply:
Actually, no.  That would be true on the AVR which is Harvard architecture (completely separate program and RAM) and you need special instructions to read flash.  ESP8266 has von Neumann, unified address space, but has alignment requirements in certain regions.

pgm_read_*() is always safe and just ensures reads are done on a 32-bit boundary (by shifting/masking after the fact).

Reply:
Well, maybe I was too influenced by AVR. Also confused by `__FlashStringHelper` and `FPSTR` that only `String` can understand. Thanks for the details!
Then in the general case when performance is not in the loop we can always use `pgm_*`.
I should have read `memcpy_P()`'s code before :)